pkg/controller/registry/resolver/cache/predicates.go

```diff
@@ -329,3 +330,41 @@ func (c countingPredicate) String() string {
 func CountingPredicate(p Predicate, n *int) Predicate {
 	return countingPredicate{p: p, n: n}
 }
+
+type evaluatorPredicate struct {
+	evaluator constraints.Evaluator
+	rule      string
+	message   string
```

Comment:
Maybe I'm being naive, but I'm a little surprised to see constraint-specific logic on the OLM side of this. I was imagining that we'd define a generic `Constraint` interface in `o-f/api` that `Cel`, `All`, `Any`, `None`, etc. would all implement.

That way, all OLM would need to do would be unmarshal the constraints and then use that generic interface to evaluate them.  What am I missing?

Reply:
This is predicate information. The `rule` and `message` are there to surface those information to the resolver during failure with `.String()` function. It is not the same as `Constraint` struct in `api` repo.